webviz_subsurface/plugins/_well_completions.py

```diff
@@ -483,10 +570,103 @@ def extract_wells(
+                zonedict["color"] = next(
+                    color_iterator
+                )  # theme colors only applied on leaves
+            else:
+                zonedict["color"] = "#808080"  # grey
+        if "subzones" in zonedict:
+            zonedict["subzones"] = add_colors_to_stratigraphy(
+                zonedict["subzones"], zone_color_mapping, color_iterator
+            )
+    return stratigraphy
+
+
+def filter_valid_nodes(
+    stratigraphy: List[Dict[str, Any]], valid_zone_names: list
+) -> Tuple[List, List]:
+    """Returns the stratigraphy tree with only valid nodes.
+    A node is considered valid if it self or one of it's subzones are in the
+    valid zone names list (passed from the lyr file)
+
+    The function recursively parses the tree to add valid nodes.
+    """
+    output = []
+    remaining_valid_zones = valid_zone_names
+    for zonedict in stratigraphy:
+        if "subzones" in zonedict:
+            zonedict["subzones"], remaining_valid_zones = filter_valid_nodes(
+                zonedict["subzones"], remaining_valid_zones
+            )
+        if zonedict["name"] in remaining_valid_zones:
+            if "subzones" in zonedict and not zonedict["subzones"]:
+                zonedict.pop("subzones")
+            output.append(zonedict)
+            remaining_valid_zones = [
+                zone for zone in remaining_valid_zones if zone != zonedict["name"]
+            ]  # remove zone name from valid zones if it is found in the stratigraphy
+        elif "subzones" in zonedict and zonedict["subzones"]:
+            output.append(zonedict)
+
+    return output, remaining_valid_zones
+
+
+def extract_stratigraphy(
+    layer_zone_mapping: Dict[int, str],
+    stratigraphy: Optional[List[Dict[str, Any]]],
+    zone_color_mapping: Optional[Dict[str, str]],
+    theme_colors: list,
+) -> List[Dict[str, Any]]:
     """Returns the stratigraphy part of the data set"""
-    color_iterator = itertools.cycle(colors)
-    return [
-        {"name": zone, "color": next(color_iterator)}
-        for zone in dict.fromkeys(layer_zone_mapping.values())
-    ]
+    color_iterator = itertools.cycle(theme_colors)
+
+    if stratigraphy is None:
+        return [
+            {
+                "name": zone,
+                "color": zone_color_mapping[zone]
+                if zone_color_mapping is not None and zone in zone_color_mapping
+                else next(color_iterator),
+            }
+            for zone in dict.fromkeys(layer_zone_mapping.values())
+        ]
+
+    # If stratigraphy is not None the following is done:
+    stratigraphy, remaining_valid_zones = filter_valid_nodes(
+        stratigraphy, list(set(layer_zone_mapping.values()))
+    )
+
+    if remaining_valid_zones:
+        raise ValueError(
+            f"""
+            The following zones are defined in the zone->layer mapping, but not in the
+            stratigraphy: {remaining_valid_zones}
+            """
```

Comment:
```suggestion
            "The following zones are defined in the zone → layer mapping, "
            f"but not in the stratigraphy: {remaining_valid_zones}"
```
Otherwise the error will be written with a lot of white space between "but not in the" and "stratigraphy [...]".